keps/sig-node/draft-20180702-pod-restartpolicy-to-restart-whole-pod.md

```diff
@@ -0,0 +1,125 @@
+However, there are scenarios (some documented in [this issue][issue]) where the many containers in the pod (including init containers) might be interlinked or inter-dependent in such a way as to require closer co-ordination when any one of its containers are restarted.
+
+### Goals
+
+Make it possible to declaratively specify that the whole pod (including the init containers) should be restarted if any container is going to be restarted.
+
+This can simplify many scenarios requiring close co-ordination of containers of a pod during individual container restart.
+
+For example, if init containers are used to verify, initialize and if necessary restore from backup the data for some persistent services, then restarting the pod when any of its regular containers crash or restart would make sure that the data is always consistent and ready before the regular containers are started or restarted. This can make the services more self-managed. It can also enhance the reach of the init containers into many other use-cases where they cannot be used right now.
+
+Many other use-cases are documented in the [upstream issue][issue].
+
+### Non-Goals
+
+It is not a goal of this proposal to address all possible ways to co-ordinate between the containers of the same pod.
+The only goal is to address the co-ordination of restarting of containers of the same pod.
+
+## Proposal
+
+Introduce a new value (`AlwaysPod`) for `restartPolicy` which works almost exactly like `Always` except that whenever the `Always` restart policy would have restarted any one of the containers of a pod, the `AlwaysPod` restart policy would restart the whole pod (including re-executing it's init containers).
+
+### Implementation Details/Notes/Constraints
+
+#### Recreate Pod Sandbox
+
+We can trigger the pod restart by triggering the recreation of the [pod sandbox][sandbox]. A work-in-progress implementation of this approach can be found [in this pull request][pr].
+
+##### Pros
+  * This approach as the benefit of having only a small amount of change to the existing code-base.
+  * It also leverages the existing mechanisms for pod restart such as during changes to the pod specification.
+##### Cons
+  * This approach is sub-optimal in that it discards the existing sandbox and the associated container instances.
+
+#### Re-use Pod Sandbox
+
+We can think of re-using the existing pod sandbox and merely restart the existing container instances in the right order. This is made a bit more challenging by the fact that the kublet currently optimizes by [pruning][prune] the container instances of successfully executed init containers.
+
+##### Pros
+  * This approach is more optimal.
+##### Cons
+  * This approach is more complex.
+  * The impact on the rest of the kubelet behaviour also might be larger.
+
+### Risks and Mitigations
+
+The `restartPolicy` or `AlwaysPod` would be a new value for an existing field in the pod specefication. So, the question of backward compatibility may not apply.
```

Comment:
This would not be backwards compatible for old kubelets, so we can’t change restartPolicy in v1 of the api. Something like “restartAffects: Pod” or a lifecycle hook would be necessary.  It’s really orthogonal to restart policy and could be used with OnFailure or Always.